[DNM]roachtest: mixed version upgrade test for schema changes

spaskob opened the pull request:
This PR is for illustration puposes only. It shows the differeces between this test
and `acceptance/version-upgrade` so that we can glean what th right abstraction are
to be shared by these tests. Also the test has some in-efficiencies: for example
the new binary is copied to all nodes every time a single node  is upgraded.

Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47237)
<!-- Reviewable:end -->

tbg commented:
Thanks! This gave me some ideas which I enshrined in this totally untested branch:

https://github.com/tbg/cockroach/pull/new/version-rand-schema

I hope the idea is clear - we stop trying to use the "feature tests" which are just too opinionated to be useful (and think should be relegated to a lesser role, i.e. less built into the version testing primitive).

Instead, we make you a helper that is well-suited to what you want, which is starting a background operation as a step at one point and stopping it again as a later step (checking for errors).

Let me know if you're confused about anything going on in that PR or if you think I've missed a requirement that is still awkward.

And thanks for working through this with me, I feel like we're about a step or two away from coming up with the primitives that allow everyone to write powerful mixed-version tests without really doing a lot of work. That will be a giant win.

Commit:
workload/schemachange: add verbosity levels and improve stats

This workload prints unsuccessful  attempts to produce a random
operation as `NOOP: <reason for failure>`. These are not that
interesting since they don't report failures of schema changes.
But they can still be useful for debugging the workload so we move
them to verbosity level 2. In addition we count separately the stats
for commit and rollback errors in the workload histogram.

Release note: none.

Commit:
roachtest: add new mixed version schema change test

This test performs a rolling upgrade, downgrade and upgrade again
from predecessor version of `HEAD` to `HEAD` while running the
`schemachange` workload in between each upgrade/downgrade of a node.

Closes #47152.

Release note: none
